src/hotspot/share/gc/parallel/psOldGen.cpp

```diff
@@ -120,6 +120,15 @@ void PSOldGen::initialize_performance_counters(const char* perf_data_name, int l
                                       _object_space, _gen_counters);
 }
 
+HeapWord* PSOldGen::expand_and_allocate(size_t word_size) {
+  assert(SafepointSynchronize::is_at_safepoint(), "precondition");
+  assert(Thread::current()->is_VM_thread(), "precondition");
+  if (object_space()->needs_expand(word_size)) {
+    expand(word_size*HeapWordSize);
+  }
+  return cas_allocate_noexpand(word_size);
```

Comment:
This implies a race and `cas_allocate_noexpand()` can fail prematurely due to the race. Is it an acceptable case?

Reply:
Can you elaborate what are racing here?

Reply:
Okay, let me ask in a different way: why use `cas_allocate_noexpand()` here, which implies there may be other threads (e.g. GC threads) that may perform allocation concurrently at this moment?

Reply:
There is no non-cas version API for allocate. Maybe I can add a comment to explain, sth like, "Reuse the CAS API even though this is VM thread in safepoint. This method is not invoked repeated, so the CAS overhead should be negligible."?

Reply:
Thank you for the explanation. A comment will be good.